### ISSUE TITLE
bug(MatFormFieldControl): Custom MatFormFieldControl does not shows errors on form submit unless touched

### ISSUE THREAD

**Issue by marekwalach** [angular/components#23387]
#### Reproduction

Steps to reproduce:
1. Follow documentation https://material.angular.io/guide/creating-a-custom-form-field-control to create custom MatFormFieldControl with ControlValueAccessor
2. Use it in Reactive Form (with some other simple fields)
3. Try to save form without touching custom field.
4. 
StackBlitz (forked from documentation example with added form element and submit button): [stackblitz](https://stackblitz.com/edit/angular-yt6hj2?file=src/app/form-field-custom-control-example.ts)
 
#### Expected Behavior

Form is invalid from the start and on submit all custom MatFormFieldControls are marked as invalid even if not touched.

#### Actual Behavior

When submitting form all standard fields will be marked as invalid (red) but custom one will not. If you write something into this custom field and then remove this value (making it dirty and touched) then it will correctly be shown as invalid. What is more whole form is VALID if this field is not touched - after you it is dirty then whole form is correctly INVALID.

#### Environment

- Angular: 12.2.1
- CDK/Material: 12.2.1
- Browser(s): all
- Operating System (e.g. Windows, macOS, Ubuntu): all

**Comment by BojanKogoj** [comment]
Your error state matchers are different.

What you have is

```ts
get errorState(): boolean {
  return this.ngControl.errors !== null && !!this.ngControl.touched;
}
```

But [Material will check for submitted](https://github.com/angular/components/blob/babfb215c2812f66aa324f3c4b09b85c278b4943/src/material/core/error/error-options.ts#L23) as well

```ts
isErrorState(control: FormControl | null, form: FormGroupDirective | NgForm | null): boolean {
  return !!(control && control.invalid && (control.touched || (form && form.submitted)));
}
```

I assume that guide doesn't have `.submitted` as well for simplicity.

**Comment by marekwalach** [comment]
@BojanKogoj I updated SB with your code but it changes nothing.

**Comment by BojanKogoj** [comment]
tldr; I'm not sure how to fix it

Your custom ErrorStateMatcher is still not being used.
You need to use [mixinErrorState](https://github.com/angular/components/blob/fbb80fcaad09ea22cb75c1a67382e98bb29ab631/src/material/core/common-behaviors/error-state.ts#L47) to use ErrorStateMatcher this way. This function will call `matcher.isErrorState(...)` and set errorState.

However I wasn't able to inject NgForm to get this to work in your example.

**Comment by BojanKogoj** [comment]
Your CustomStateMatcher isn't being used. Material uses [mixinErrorState](https://github.com/angular/components/blob/fbb80fcaad09ea22cb75c1a67382e98bb29ab631/src/material/core/common-behaviors/error-state.ts#L47) which calls isErrorState.

I [modified your example](https://stackblitz.com/edit/angular-yt6hj2-25bafs?file=src/app/form-field-custom-control-example.ts) , changed errorState

```ts
  get errorState(): boolean {
    const parent = this.parentForm || this.parentFormGroup;
    return !!this.parts.invalid && (!!this.parts.touched || parent.submitted);
  }
```

And added injected FormGroup and NgForm.

```ts
    @Optional() private parentForm: NgForm,
    @Optional() private parentFormGroup: FormGroupDirective
```

**Comment by marekwalach** [comment]
@BojanKogoj that work with error state which "marks field red". But whole form is still VALID without writing anything into this field.

**Comment by BojanKogoj** [comment]
Same as https://github.com/angular/components/issues/23352
Not sure how to solve this right now.

**Comment by viniciusschuelter** [comment]
@marekwalach I'm missing something, maybe it's an emit of the value when the writeValue is started, I believe that implementing ngAfterViewInit with an onChange would solve this problem.

`ngAfterViewInit(): void {
    this.onChange(this.value);
  }
`

Can you confirm if it works?

**Comment by marekwalach** [comment]
@viniciusschuelter it changes nothing unfortunately.

**Comment by Bezmehrabi** [comment]
There are some error triggers that we can't subscribe to (e.g. parent form submissions).
so you just have to check for error state update on DoCheck.
reference:
https://github.com/angular/components/blob/main/src/material/input/input.ts#L350

I will make a PR for it to make the docs more clear.
